[quant][graphmode] Insert Observers for dynamic LSTM

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35916 [quant][graphmode] Add quantize_per_tensor.tensors
* **#35894 [quant][graphmode] Insert Observers for dynamic LSTM**
* #35893 [quant][graphmode] Add new tensorlist observer for LSTM

Summary: Insert new TensorListObserver only for weight input of dynamic LSTM
This is because we are currently not observing the activation inputs in graph mode.
Activation tensors are dynamically quantized within the aten::qlinear_dynamic op

Test Plan:
python test/quantization/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20830387](https://our.internmc.facebook.com/intern/diff/D20830387)

supriyar committed:
[quant][graphmode] Insert Observers for dynamic LSTM
Summary: Insert new TensorListObserver only for weight input of dynamic LSTM
This is because we are currently not observing the activation inputs in graph mode.
Activation tensors are dynamically quantized within the aten::qlinear_dynamic op

Test Plan:
python test/quantization/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -745,17 +745,17 @@ bool isBiasOfConvOrLinear(Value* v) {
 }
 
-bool isWeightOfConvOrLinear(Value* v) {
+bool isWeightOfFunction(Value* v) {
```
I think we can just use `isWeight`

supriyar committed:
Update on "[quant][graphmode] Insert Observers for dynamic LSTM"


Summary: Insert new TensorListObserver only for weight input of dynamic LSTM
This is because we are currently not observing the activation inputs in graph mode.
Activation tensors are dynamically quantized within the aten::qlinear_dynamic op

Test Plan:
python test/quantization/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -745,17 +745,13 @@ bool isBiasOfConvOrLinear(Value* v) {
   return result;
 }
 
-bool isWeightOfConvOrLinear(Value* v) {
```
we still need this, we have other code that relies on this function

supriyar replied:
I've moved the logic inside `isWeight`. All the existing tests are passing with this change.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -782,9 +778,8 @@ bool isWeight(Module& module, Value* v) {
-Module getObserverModuleFor(Value* v, const QConfig& qconfig) {
-  return isWeightOfConvOrLinear(v) ? std::get<1>(qconfig)
-                                   : std::get<0>(qconfig);
+Module getObserverModuleFor(Module& m, Value* v, const QConfig& qconfig) {
```
can you keep two `isWeight` function for now? although this should still work since we have a valueNeedsToBeQuantized check before calling this function, but we don't really need to go through CallMethod here

jerryzh168 reviewed (changes requested):
please revert changes to `getObserverModuleFor`

supriyar committed:
Update on "[quant][graphmode] Insert Observers for dynamic LSTM"


Summary: Insert new TensorListObserver only for weight input of dynamic LSTM
This is because we are currently not observing the activation inputs in graph mode.
Activation tensors are dynamically quantized within the aten::qlinear_dynamic op

Test Plan:
python test/quantization/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
